### PR TITLE
Add IOManager and OutputBuffer

### PR DESCRIPTION
**Pull request by davidbeckingsale**
This adds three ostreams:

- `umpire::log`
- `umpire::replay`
- `umpire::error`

By default, the log stream is set up to print messages to both `std::cout` and to a  file named `<base_dir>/umpire.<uid>.log`, where `base_dir` is a specified base directory. and `uid` is a monotonically increasing random ID to ensure files are never overwritten if running multiple times with the same `base_dir`.

Both `Logger` and `Replay` have been updated to use the appropriate ostreams.

TODO:

- [x] Tests
- [x] Add MPI rank to output filename
- [x] Documentation

### CODE REVIEW

File: src/umpire/Replay.cpp [hunk 1]
```diff
@@ -43,7 +44,7 @@ void Replay::logMessage( const std::string& message )
   if ( !replayEnabled )
     return;   /* short-circuit */
 
-  std::cout << message;
+  umpire::replay << message;
```
Comment (mcfadden8) [inline src/umpire/Replay.cpp]: @davidbeckingsale, this file actually needs to put stuff to standard output.  The output here is from the program to the user.

File: src/umpire/Umpire.hpp [hunk 2]
```diff
@@ -25,6 +28,19 @@
 
 namespace umpire {
 
+inline void
+initialize()
```
Comment (davidbeckingsale) [inline src/umpire/Umpire.hpp]: Do we want to initialize like this?

File: src/umpire/util/IOManager.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  s_root_io_dir = dir;
+}
+
+void
+IOManager::initialize()
+{
+  if (!s_initialized) {
+    auto output_dir = std::getenv("UMPIRE_OUTPUT_DIR");
```
Comment (davidbeckingsale) [inline src/umpire/util/IOManager.cpp]: @mcfadden8 - added environment variables.

File: src/umpire/util/MPI.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+int MPI::s_rank = -1;
+int MPI::s_world_size = -1;
+bool MPI::s_initialized = false;
+
+void 
+MPI::initialize()
+{
+  if (!s_initialized) {
+#if !defined(UMPIRE_ENABLE_MPI)
```
Comment (davidbeckingsale) [inline src/umpire/util/MPI.cpp]: MPI support here!

Comment (mcfadden8) [reply src/umpire/util/MPI.cpp]: Does this build without including MPI?  Or are we always going to build with MPI?

Comment (davidbeckingsale) [reply src/umpire/util/MPI.cpp]: Of course. If MPI isn't enabled, we "fake" it and return 0 for rank, and 1 for size.

**Comment by davidbeckingsale** [comment]
@mcfadden8 @jdahm - what are you thoughts on how to test output?

**Comment by jdahm** [comment]
- Unit test: write something to an output stream, then read that stream back in and check that the output matches.
- Integration test: trigger a log, then test that it gets created correctly.

**Comment by davidbeckingsale** [comment]
Adding tests, example output:

```
22: [--------]
22: [RUN     ] Checking for "testing log stream" in stdout
22: [      OK] Found "testing log stream" in stdout
22: [RUN     ] Checking for "testing error stream" in stderr
22: [      OK] Found "testing error stream" in stderr
22: [RUN     ] Checking umpire_io_tests.0.0.log exists
22: [      OK] umpire_io_tests.0.0.log exists
22: [RUN     ] Checking umpire_io_tests.0.0.replay exists
22: [      OK] umpire_io_tests.0.0.replay exists
22: [RUN     ] Checking for "testing log stream" in umpire_io_tests.0.0.log
22: [      OK] Found "testing log stream" in umpire_io_tests.0.0.log
22: [RUN     ] Checking for "testing replay stream" in umpire_io_tests.0.0.replay
22: [      OK] Found "testing replay stream" in umpire_io_tests.0.0.replay
22: [--------]
```

File: tests/integration/io/iomanager_tests_runner.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    'END': '\033[0m',
+}
+
+
+errors = 0
+
+def check_output(name, file_object, expected):
+    global errors
+
+    print("{GREEN}[RUN     ]{END} Checking for \"{expected}\" in {name}".format(name=name, expected=expected, **formatters))
```
Comment (mcfadden8) [inline tests/integration/io/iomanager_tests_runner.py]:
```suggestion
    print("{BLUE}[RUN     ]{END} Checking for \"{expected}\" in {name}".format(name=name, expected=expected, **formatters))
```

File: tests/integration/io/iomanager_tests_runner.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    contents = file_object.readline().rstrip()
+    if (contents != expected):
+        print("{RED}[   ERROR]{END} Got {contents}".format(contents=contents, expected=expected, **formatters))
+        errors = errors + 1
+    else:
+        print("{GREEN}[      OK]{END} Found \"{expected}\" in {name}".format(name=name, expected=expected, **formatters))
```
Comment (mcfadden8) [inline tests/integration/io/iomanager_tests_runner.py]:
```suggestion
        print("{BLUE}[      OK]{END} Found \"{expected}\" in {name}".format(name=name, expected=expected, **formatters))
```

File: tests/integration/io/iomanager_tests_runner.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    import os
+
+    global errors
+
+    print("{GREEN}[RUN     ]{END} Checking {myfile} exists".format(myfile=filename, **formatters))
+    if (not os.path.isfile(filename)):
+        print("{RED}[   ERROR]{END} {myfile} not found".format(myfile=filename, **formatters))
+        errors += errors + 1
+    else:
+        print("{GREEN}[      OK]{END} {myfile} exists".format(myfile=filename, **formatters))
```
Comment (mcfadden8) [inline tests/integration/io/iomanager_tests_runner.py]:
```suggestion
        print("{BLUE}[      OK]{END} {myfile} exists".format(myfile=filename, **formatters))
```

File: tests/integration/io/iomanager_tests_runner.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    with open(replay_filename) as replay_file:
+        check_output(replay_filename, replay_file, 'testing replay stream')
+
+
+
+if __name__ == '__main__':
+    import sys
+
+    print("{GREEN}[--------]{END}".format(**formatters))
```
Comment (mcfadden8) [inline tests/integration/io/iomanager_tests_runner.py]:
```suggestion
    print("{BLUE}[--------]{END}".format(**formatters))
```

File: tests/integration/io/iomanager_tests_runner.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    run_io_test({'UMPIRE_OUTPUT_BASENAME' : 'umpire_io_tests'}, 0)
+    run_io_test({'UMPIRE_OUTPUT_DIR': './io_test_dir', 'UMPIRE_OUTPUT_BASENAME' : 'umpire_io_tests'}, 0)
+    run_io_test({'UMPIRE_OUTPUT_DIR': './io_test_dir', 'UMPIRE_OUTPUT_BASENAME' : 'umpire_io_tests'}, 1)
+    print("{GREEN}[--------]{END}".format(**formatters))
```
Comment (mcfadden8) [inline tests/integration/io/iomanager_tests_runner.py]:
```suggestion
    print("{BLUE}[--------]{END}".format(**formatters))
```

File: tests/integration/io/iomanager_tests_runner.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+#
+# This file is part of Umpire.
+#
+# For details, see https://github.com/LLNL/Umpire
+# Please also see the LICENSE file for MIT license.
+##############################################################################
+
+formatters = {
+    'RED': '\033[91m',
+    'GREEN': '\033[92m',
```
Comment (mcfadden8) [inline tests/integration/io/iomanager_tests_runner.py]:
```suggestion
    'BLUE': '\033[94m',
```

**Comment by davidbeckingsale** [comment]
@mcfadden8 - I made the changes you suggested, thanks!

**Pull request merged**